GPflow/conditionals.py

```diff
@@ -2,246 +2,106 @@
     A = tf.matrix_triangular_solve(Lm, Kmn, lower=True)
-    B = tf.matrix_triangular_solve(tf.transpose(Lm), A, lower=False) # B is Kmm^{-1} Kmn
 
-    #construct the mean and variance of q(f*)
-    fmean = tf.matmul(tf.transpose(B), q_mu)
+    #compute the covariance due to the conditioning
     if full_cov:
         fvar = kern.K(Xnew) - tf.matmul(tf.transpose(A), A)
         fvar = tf.expand_dims(fvar, 2)
     else:
         fvar = kern.Kdiag(Xnew) - tf.reduce_sum(tf.square(A), 0)
         fvar = tf.expand_dims(fvar, 1)
-    projected_var = []
-    for d in range(num_columns):
-        if q_sqrt.get_shape().ndims==2:
-            LTB = B*q_sqrt[:,d:d+1]
-        elif q_sqrt.get_shape().ndims==3:
-            L = tf.user_ops.triangle(q_sqrt[:,:,d], 'lower')
-            LTB = tf.matmul(tf.transpose(L), B)
-        else:
-            raise ValueError, "Bad dimension for q_sqrt: %s"%str(q_sqrt.get_shape().ndims)
-        if full_cov:
-            projected_var.append(tf.matmul(tf.transpose(LTB),LTB))
-        else:
-            projected_var.append(tf.reduce_sum(tf.square(LTB),0))
-    fvar = fvar + tf.transpose(tf.pack(projected_var))
-
-    return fmean, fvar
-
-
-def gaussian_gp_predict_whitened(Xnew, X, kern, q_mu, q_sqrt, num_columns, full_cov=False):
-    """
-    Given an (approximate) posterior (via q_mu, q_sqrt) to the GP at the points
-    X, produce the mean and variance of the GP at the points Xnew.
-    Additionally, the GP has been centered (whitened) so that 
-        p(v) = N( 0, I)
-        f = L v
-    thus
-        p(f) = N(0, LL^T) = N(0, K).
-
-    We assume K independent GPs, represented by the columns of q_mu (and the
-    last ax of q_sqrt).  q_mu and q_sqrt are variational posteriors for v, So
-        q(v[:,i]) = N( q_mu[:,i], diag(q_sqrt[:,i]**2)
-        q(f[:,i]) = N (L q_mu[:,i],  L diag(q_sqrt**2) L^T)
-    or
-        q(f[:,i]) = N (L q_mu,  L [W W^T] L^T)
-    where W is the lower triangle of q_sqrt[:,:,i].
-
-    This function computes the Gaussian integral
-        q(f*) = \int p(f*|(f=Lv))q(v) df.
-
-    Xnew is a data matrix, size N x D
-    X are data points, size M x D
-    q_mu are variational means, size M x K
-    q_sqrt are variational standard-deviations or Cholesky matrices,, size M x K or M x M x K
-
-    Note (and TODO):
-        At the moment, num_columns only gets used for the q_sqrt.ndim==3 case,
-        and it tells use the value of q_sqrt.shape()[2]. We need to find a way
-        to get this from the tf graph.
-
-
-    See also:
-        gp_predict_whitened -- where there is no uncertainty in V
-        gaussian_gp_predict -- same without the whitening
-
-    """
-
-    #compute kernel stuff
-    num_data = tf.shape(X)[0]
-    Kmn = kern.K(X, Xnew)
-    Kmm = kern.K(X) + eye(num_data)*1e-6
-    Lm = tf.cholesky(Kmm)
 
-    #this is O(N M^2)
-    A = tf.matrix_triangular_solve(Lm, Kmn, lower=True)
+    #another backsubstitution in the whitened case
+    if not whiten:
```

Comment:
Yes, good spot.

On 30 March 2016 at 13:14, Alexander G. de G. Matthews <
notifications@github.com> wrote:

> In GPflow/conditionals.py
> <https://github.com/GPflow/GPflow/pull/27#discussion_r57877820>:
>
> >
> > -    #this is O(N M^2)
> > -    A = tf.matrix_triangular_solve(Lm, Kmn, lower=True)
> > +    #another backsubstitution in the whitened case
> > +    if not whiten:
>
> Shouldn't the comment here say "in the unwhitened case"?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/GPflow/GPflow/pull/27/files/6a6e90625bc7dac0c02e6a8d3a7395424bb5076e..2546cd8b659eeaeb4dbff07d4880b59a883596a0#r57877820>
>